sound/hybrid_font.h

```diff
@@ -230,16 +231,45 @@ class HybridFont : public SaberBase, public Looper {
     if (loop) hum_player_->PlayLoop(loop);
   }
 
+  RefPtr<BufferedWavPlayer> GetOrFreeWavPlayer(Effect* e) {
+    free_player_ = GetFreeWavPlayer();
+    if (!free_player_) {
+      STDOUT.println("Out of WAV players! Getting more...");
+      float highest_progress = 0;
+      int player_to_restart = -1;
+      // check wavplayers, rule out hum and smoothswings then kill oldest one 
+      for (size_t i = 0; i < NELEM(wav_players); i++) {
+        if (wav_players[i].isPlaying() && 
+          wav_players[i].current_file_id().GetEffect() &&
```

Comment:
It's not safe to just kill *any* player, since we don't know what they are used for in this code.
Basically, if you're going to consider all sounds killable, you also need to review all code that plays sound and make sure that any sound that we don't want to to be killed has a reference so that it doesn't get killed.

Also, if we end up in a situation where all the sounds that are playing are using different effects, then I don't think we should kill any of them. Only effects that play at least two sounds should be considered killable IMHO.

Reply:
So here's example output from a test. It prints out "file not referenced" only if it satisfies the if statement checking for that. 
hum, smoothswings and track are referenced and therefore not subject to being killed.
```
21:50:02.717 -> Out of WAV players! Getting more...
21:50:02.717 -> *************** Loop i = 0 wav_player has aa_Juansith/Red_Five/humm/humm02.wav
21:50:02.717 -> *************** Loop i = 1 wav_player has aa_Juansith/Red_Five/tracks/Xwing.wav
21:50:02.717 -> *************** Loop i = 2 wav_player has aa_Juansith/Red_Five/swingh/swingh01.wav
21:50:02.717 -> *************** Loop i = 3 wav_player has aa_Juansith/Red_Five/swingl/swingl01.wav
21:50:02.717 -> *************** Loop i = 4 wav_player has aa_Juansith/Red_Five/blst/blst08.wav
21:50:02.717 -> *************** file is not referenced
21:50:02.717 -> *************** oldest so far, setting it as the player_to_restart =4
21:50:02.717 -> *************** Loop i = 5 wav_player has aa_Juansith/Red_Five/blst/blst01.wav
21:50:02.717 -> *************** file is not referenced
21:50:02.717 -> *************** oldest so far, setting it as the player_to_restart =5
21:50:02.717 -> *************** Loop i = 6 wav_player has aa_Juansith/Red_Five/blst/blst01.wav
21:50:02.717 -> *************** file is not referenced
21:50:02.717 -> Stopping wav_player playing aa_Juansith/Red_Five/blst/blst01.wav
```
This is working, no?

I could really use some understanding of a negative result and how to handle it.
Like what do we do if the checks fail. How do we handle reaching the end of this process without an answer?
What would we return?
Like if we're going to NOT kill the player because there's only 1 instance of that kind of sound, what's the exit strategy then since we need to go back to PlayPolyphonic and play something but we're empty handed and don't have a free player?
Lastly, would using `std::string` be a way to get which type of sound it is(clsh, blst, etc..) so as to build a check for multiple of the same kind?

Reply:
> This is working, no?

Perhaps, but it's not enough for it to work, it must also make sense.
This means that you have to be able to make a reasonable argument that explains why it works, not just try it.
The way the code is written now, this means every other piece of code that plays sounds in ProffieOS and make sure that it isn't impacted by what you're doing. Earlier, when you were only preempting players that were playing the *same* effect, that was much much much easier. Of course, that doesn't always work, but it's *easier*.

> Like what do we do if the checks fail. How do we handle reaching the end of this process without an answer?
> What would we return?

Just like GetFreeWavPlayer, we would return a null refptr if we fail to find an available player.

> Like if we're going to NOT kill the player because there's only 1 instance of that kind of sound, what's the exit strategy > then since we need to go back to PlayPolyphonic and play something but we're empty handed and don't have a free >player?

In this case PlayPolyphonic() will not play anything, just like it does today if GetFreeWavPlayer fails to return a player.

> Lastly, would using std::string be a way to get which type of sound it is(clsh, blst, etc..) so as to build a check for multiple of the same kind?
.
Most of the time, you would count multiples by using std::map<>, iterating over each player and then doing something like map[player[i].getEffect()]++. Unfortunately, std::map allocates memory, and it's not a good idea to use it in proffieos.
Instead we probably have to use a double loop, something like:

   for ( i  = 0 .. players ) {
       Effect e = player[i].geteffect();
       count = 0;
       for ( j = 0 .. players ) {
          if (player[j].geteffect() == e) count++;
       }
       if (count > 1) {
          // do stuff
       }
   }

This takes a little more cpu, but since it won't run that often it's not that big of a deal.

Reply:
> Effect e = player[i].geteffect(); 

I think I interpreted your shorthand for the most part, but this one's giving me
`error: 'class BufferedWavPlayer' has no member named 'GetEffect'`
I tried adding `current_file_id` like elsewhere, I also tried asterisk `Effect*`, `const Effect*`
to no avail.

Reply:
I post too quickly and make it hard to follow. apologies.

It seems we don't need the loop in the loop? We are already checking if the sound is the effect we want in the if statement of the first for-loop. We can just increment count there once that check passes, no. 
The problem is if there's a fail, it freezes with `NULL POINTER!`
Tried:
return RefPtr<BufferedWavPlayer>(nullptr);
return RefPtr<BufferedWavPlayer>();
return RefPtr<BufferedWavPlayer>;

Reply:
SO I went another step and added saving the 2nd oldest playing sound as well. The idea is if a new effect comes in, and the loop doesn't get a player to restart, then we'd use the backup.
However, this has reached a point of diminishing returns.
All this checking and processing causes a lag when it runs no different than just waiting like the original code did.
Unless you can optimize some things to have this run quicker, I'm not sure it's worth it.

It used to be snappy a few iterations ago, before checking quantity playing. Worked well actually. 
I don't know if you've tried it in real life, but I'd say there's just no way you'd hear that ANY are being killed, much less which one. It's a frantic cacophony of effect sounds. The goal was no dead spots when spamming the button or clashing like crazy, which it satisfied.
If the concern of hearing a sound get cut off could be reconsidered, this has a nice result. I could revert back to what works safely, makes sense, and maybe be done?

Reply:
Running all of this code takes a couple of microseconds, so I don't think that's the problem.

Reply:
I think this is what I was thinking.
So wav_players are all full. We come through with an effect that's new.
Because we fail the if statement for if == e, we do not choose a player to kill.
This is where a saved_player_to_restart, which we acquired the last time around as the 2nd oldest sound, can be killed and make room for our new current effect to play.

Reply:
Ok, now I understand why it's here.
But it doesn't really work, because the "saved" variable is local and will not survive until the next call.
How about this, we add a bit to the Effect class, but instead of adding an ini file setting for it, we just set it from this function.

Basically, at the beginning of this function we would do: e->set_killable(true);
Then when we iterate over the players, we check if it's killable:

    Effect* effect = wav_players[i].GetEffect();
    if (!effect || !effect->is_killable()) continue;

Reply:
> Ok, now I understand why it's here. But it doesn't really work, because the "saved" variable is local and will not survive until the next call. 

int saved_player_to_restart_ = -1 ; is initialized outside the function and therefore not local, and should, and does work that way in tests. 


How about this, we add a bit to the Effect class, but instead of adding an ini file setting for it, we just set it from this function.
> 
> Basically, at the beginning of this function we would do: e->set_killable(true); Then when we iterate over the players, we check if it's killable:
> 
> ```
> Effect* effect = wav_players[i].GetEffect();
> if (!effect || !effect->is_killable()) continue;
> ```

Ok, however I think there's 2 different objectives at play here. The first being NOT killing an incorrect file, as you've addressed right here. I'm still unsure why more needs to be done here, as referenced sounds are indeed being ignored as we iterate through the players. Perhaps referenced is not enough "proof" it's not killable and I don't understand, and/or maybe I need to try to test with in.wavs?

But the second (and my primary) objective is trying to prevent a missed effect, a dead button... a dud.
That's what I was working on above. Killing the next viable player based on previously recorded oldest duration when a new effect is current (and therefore not triggering a kill based on itself). That code is also working as-is in tests according to a slew of printouts. 

Last I tested with the printouts though, there was new, significant delay everytime we hit "Out of WAV Players!" and GetOrFreeWavPlayer() ran. As you pointed out, we're talking milliseconds to run that bit, so something else must have crept up. Won't have hands-on test time for about a week though as of today.